sympy/assumptions/handlers/matrices.py

```diff
@@ -38,6 +38,8 @@ def MatMul(expr, assumptions):
         factor, mmul = expr.as_coeff_mmul()
         if all(ask(Q.symmetric(arg), assumptions) for arg in mmul.args):
             return True
+        if ask(Q.diagonal(expr), assumptions):
```

Comment:
(I also responded on the mailing list). The old handlers system doesn't work so well with deductions.  Particularly, it doesn't round trip the known facts. Basically, it checks the `known_facts` first, then falls back to the handlers for more complicated deductions. This is what the new satask system does better. If you implement this in sathandlers, it should work, because everything uses the SAT solver, so the known_facts are always part of any deductions. 

I don't know how difficult this will be. No facts for matrices have been added yet.